server/src/game/world.ts

```diff
@@ -13,72 +14,68 @@ export interface Coords {
 
 const DAMAGE_MULTIPLIER = 20;
 
-type WorldImportArgs = [Map, { [civID: number]: Civilization }, number, { [leaderID: number]: Leader }, number];
+type WorldImportArgs = [Map, { [civID: number]: Civilization }, number, number];
 
 export class World {
   map: Map;
   civs: { [civID: number]: Civilization };
   civsCount: number;
-  leaderPool: { [leaderID: number]: Leader };
-  updates: { (civID: number): Event }[];
+  updates: { (leader: Leader): Event }[];
   random: Random;
 
   public currentTurn: number;
 
-  constructor(args: [map: Map, civsCount: number] | WorldImportArgs) {
+  constructor(args: [map: Map, civsCount: number, civTemplateLeaders: [number, Leader][]] | WorldImportArgs) {
     this.updates = [];
 
-    if (args.length === 5) {
-      const [map, civs, civsCount, leaderPool, currentTurn] = args;
+    if (args.length === 4) {
+      const [map, civs, civsCount, currentTurn] = args;
       this.map = map;
       this.civs = civs;
       this.civsCount = civsCount;
-      this.leaderPool = leaderPool;
       this.random = new Random(map.seed);
       this.currentTurn = currentTurn;
       return;
     }
 
-    const [ map, civsCount ] = args;
+    const [ map, civsCount, civTemplateLeaders ] = args;
 
     this.random = new Random(map.seed);
     this.map = map;
 
     this.civsCount = civsCount;
     this.civs = {};
-    this.leaderPool = {};
 
     for (let civID = 0; civID < this.civsCount; civID++) {
-      this.civs[civID] = new Civilization();
-
-      this.getStartLocaltion(([settlerCoords, builderCoords, scoutCoords]) => {
-        this.addUnit(new Unit('settler', settlerCoords, civID));
-        this.addUnit(new Unit('builder', builderCoords, civID));
-        this.addUnit(new Unit('scout', scoutCoords, civID));
+      const [templateID, leader] = civTemplateLeaders[civID];
+      this.civs[civID] = new Civilization(civID, templateID);
+      const domainID = this.civs[civID].getDomainID();
+
+      this.getStartLocation(([settlerCoords, builderCoords, scoutCoords]) => {
+        this.addUnit(new Unit('settler', settlerCoords, domainID, this.civs[civID].startingKnowledge));
+        this.addUnit(new Unit('builder', builderCoords, domainID, this.civs[civID].startingKnowledge));
+        this.addUnit(new Unit('scout', scoutCoords, domainID, this.civs[civID].startingKnowledge));
       });
 
-      this.updateCivTileVisibility(civID);
+      this.setDomainLeader(domainID, leader)
```

Comment:
I would personally place this closer to the creation of the Civ (because it feels potentially dangerous to have `Civ.leaderID` be `null` for a short period of time — what if we want to look up the leader when creating the Units above, for example?)

Reply:
I think we should design things to where a Civ not having a leader is okay?